### DISCUSSION TITLE
Regulator: stealth-ultd's project

### DISCUSSION THREAD

**Discussion by stealth-ultd** [jandrassy/Regulator#1] [General]
Hi Juray,

Great to come across this well-documented DIY energy diverter. I am looking at a comparable setup for a Fronius Galvo, with the difference that in this case the utility company does not allow any power-to-grid (=zero export). This is implemented with a Fronius smartmeter (at feed-in point, modbus RTU) and with activation of Dynamic Power Reduction in the inverter DNO Editor settings ("Max. grid feed-in power: = -0.5%"). This last setting may look odd at first, but guarantees that there is always a minimum consumption taking place to avoid any export. 

What I would like to achieve is that surplus power can get diverter to a small supplementary heater (1.5 kW) while maintaining export limiting. Now I am not sure how the Regulator interacts with such a zero-export setting in the Fronius inverter.

Two initial questions about this: (1) Is there any reason to assume that the Regulator would not be compatible with zero-export, and (2) is it in any way necessary or helpful to have the secondary Fronius smartmeter (like you installed in 2019/5) for this setup?

**Comment by JAndrassy** [comment]
Hi. my project reads the Smart Meter. With zero-export there is no way to know how much power would be available. You would need to know how much power would be available without dynamic reduction. I think this information is not available anywhere. I think the MPPT configures the parameters, but there is no way to know how much would be produced with optimal parameters.

I added the secondary Smart Meter for SolarWeb data and I used it to fine tune the calculation of consumed power done in Arduino from CT readings (ElSens.ino).

**Comment by JAndrassy** [comment]
Maybe one way would be to ramp the consumption until Smart Meter reports that power is draw from grid and then regulate to some small power draw from grid.

**Comment by stealth-ultd** [comment]
Hi,
Thanks for the explanation. Interesting. It is reverse-engineering the workings of the Fronius Ohmpilot, in some way (the last works with zero-export since a recent firmware update of the data-manager, Software Version: V3.13.3-2). Curious if it follows the same approach of ramping up consumption, or if it interacts on a deeper level with the MPPT. Not sure if the Fronius SunSpec implementation gives away any clue on how this is done.

**Comment by JAndrassy** [comment]
the SunSpec has status I_STATUS_THROTTLED. So until status is THROTTLED the consumption can ramp up. (or until max power is reached).
I would love to try it, but I am very busy with work now and the season of surplus is at the end in few weeks. So it will have to wait until Spring.

**Comment by stealth-ultd** [comment]
Although little detail is available on how the Ohmpilot exactly manages this, others report indeed a gradual ramping up of the power. Will explore this a bit further.
I am using it for complementary heating in the colder seasons, so the most relevant time (autumn, spring) is starting in a bit, when solar is still abundent but not completely used. In any case, I will now first start gathering the hardware required!

**Comment by JAndrassy** [comment]
I didn't commit the code yet, but I replaced the Kemo modules with a Triac dimmer
https://robotdyn.com/ac-light-dimmer-module-1-channel-3-3v-5v-logic-ac-50-60hz-220v-110v.html
and the MCU handles the zero-crossing to phase-cutting with the Triac

**Comment by stealth-ultd** [comment]
Good to know, that makes things more simple. Does the Triac dimmer also eliminate the need for the Grove MOSFET module?

**Comment by JAndrassy** [comment]
I already forgot that I used a MOSFET :-). I had the MOSFET only for 3.3 V boards (Wemos D1 R2 and M0). I didn't use it for AVR boards (Uno (WiFi) and Mega). Same applies for the I2C ADC.

to other hardware:

I had the I2C ADC for esp8266 (D1 R2) because it has only one and not very good analog port.
And with SAMD (M0) I didn't solve analogRead from the CT yet. The readings have some inertia. Plotted waves are asymmetric.

The Chinese relays don't last long. One, maybe two seasons. First they start to generate strong EMI which disturbs the Ethernet chip. Finally they stop switching. I plan to replace them with DIN rail relays with 230 V coil switched by the new Grove 230 V sold state relay modules.

I have [an ATmega1284p board](https://github.com/budvar10/Arduino-ATmega1284P/blob/master/README.md) ready to use with Regulator but no time to put it in and test. I want to test the Triac control with AVR and revisit AVR analogRead. If the AVR doesn't turn out fantastic, then I will return to SAMD, but not to M0. I plan to buy the Arduino MKR Connector Carrier (Grove compatible) and MKR ETH shield for my MKR Zero. The configuration will better fit in the box and I hope it will be less sensitive to EMI with ground lines not going over the Eth Shield like with Grove Base Shield.

**Comment by stealth-ultd** [comment]
That’s quite some modification going on in the background! Some thoughts before updating my component list :-)

The Chinese relays failing with lots of EMI likely points to sparking (bad contact). Apart from maybe suspicious component quality, one other source could be low activation voltage / current. The relays have an operating voltage of 4.75V~5.25V at 185mA . Especially when driven with 3.3V boards the combination of voltage / current available may just be not enough to properly trigger them (two of them: 0.4 A, which is quite a draw on those boards). That had made me think of using a "30A 1-Channel Relay Module with Optocoupler H/L Level Triger”. They can be independently powered with 9 or 12 V directly from the main power supply, while the signal (3.3 or 5V) comes from the MCU. Have you maybe thought of that option too before considering the solid state relays?

I am curious how well the CT reading works at higher powers, as the CT is “only" rated at 5A and has to operate at nearly 9A (2kW). It may experience saturation of the (ferrite) core and give a distorted waveform (harmonics). I don’t know if this explains some of the CT readings at the SAMD (M0) or that something else is at play there.

EMI may in general be a plague due to all the switching electronics (SMPS, PMW, …), so I was planning to add (larger) ferrite cores to filter EMI on all the cabling getting in and out (power, ethernet, etc.) and where necessary on the internal connections.

Also I have been wondering about the connection with the inverter – I was looking at ModBus RTU, rather than TCP. We have the Fronius smartmeter already connected over RTU and extending it to the regulator may be easy – and less prone to possible communications errors from all the other “talk” that is going on over TCP (router, laptops, etc.) or EMI. I found some small Modbus RTU controllers that may do the trick.

And I am still in doubt about the Robotdyn triac dimmer versus the Kemo 150/028N module. The fist being inexpensive, but maybe underrated (or: overstated!) with risk of blowing out / the second supposedly more solid but significantly larger and expensive. Not sure where to go here. (In any case, both benefit from EMI chokes, the 028N has an extended filter design in the manual included.)

The Arduino MKR looks like a good choice indeed, especially with its small form-factor. Nice suggestion, I consider jumping on that one immediately. I guess this has no immediate effects on the sketch, as that all should be compatible?

**Comment by JAndrassy** [comment]
all boards I used in the project have a regulator and the power source is good. The planed MKR Connector Carrier would be even better. The Grove modules get only the signal from MCU, the power is from regulator in my setup.

the Grove CT is to 10 A. the datasheet has wrong calculation of ratio. For SAMD I measured only half of the current (see the dashed line in 'schematics').

With Smart Meter you can't use Modbus RTU because the inverter is Modbus RTU master.

In some moment of understanding I realized how unnecessary complicated is to go over the Kemo modules. First PWM to digital pot and then the regulation module controlde by the pot. The direct control of Triac works fantastic. It is very exact compared to commanding the modules by PWM. The formula to calculate the period for available power is simpler and cleaner. I installed the Triac in Jun so it worked all Summer.

Yes the MKR Zero is compatible with the Regulator sketch.

**Comment by JAndrassy** [comment]
I made the commit of the code with dimmer module support. and some development utility sources

**Comment by stealth-ultd** [comment]
This is very helpful - thanks.

You convinced me about direct control via the Robotdyn triac, will surely include that ;-)

Regarding Modbus RTU – I had missed that. Indeed, no other option than TCP.

As for the relays, I consider to go solid state option completely, for instance with DIN-rail mounted TRA23D25L (they include heatsink). We’ll see how this works.

Hardware list now getting complete.

**Comment by JAndrassy** [comment]
I think the TRA23D25L are based on a Triac.

a 230 V SSR could be used instead of the Triac module for phase cutting.

**Comment by JAndrassy** [comment]
Before the Kemo modules I had power regulation box controlled by PWM signal. It was based on a Triac and some MCU. I didn't understand phase cutting then. The cooler in the box was heating-up to very hot (now I know it is normal) and it didn't get to full power. Full on it enabled 1950 W but with direct connection the heater consumed 2050 W. So I implemented a 'bypass' relay for full power.
The Kemo module didn't heat-up, it was only warm, but it too doesn't allow full power.

Now with direct control of the Triac I can regulate from 10 W to 2050 W, but the cooler is very hot at higher regulated power  and on full power too. So I still use the 'bypass' for full power and my current algorithm stays in full power even if there is no surplus, if the Fronius battery is over 95% SoC. 

The Triac based SSR you linked will get hot too. And this SSR can be used to regulate the power with phase cutting and there is no point to use one as a 'bypass'. The specs table says they have 'zero-crossing'. I wonder if they turn of on zero-crossing like a plain Triac. If not, then the MCU would need to hold the pin HIGH instead of sending a short pulse to Triac.

My Triac journey started here https://arduino.stackexchange.com/questions/63631/fan-regulator-control-from-arduino/63674#63674

**Comment by stealth-ultd** [comment]
The TRA23D25L is based on a Triac. What attracted me is that it is compact, DIN-rail mounted and has a heatsink attached. Reading more about SSR’s during these last days show that they indeed may run hot like you mention – maybe more than I hoped for. We’ll see.

This specific SSR has a zero-crossing detection circuit. So it will turn on only at the ZC point. As long as the input is HIGH, the gate will remain triggered, and when the input goes LOW, it will switch off at ZC. To keep it on, the pin needs to be held HIGH.

From what I now understand, this is not the best type of SSR for phase cutting. A ‘random firing’ SSR would be better here, because it can be triggered at any desired point of the wave (but will wait for ZC to switch 'neatly' off). Among the many, this article (https://hackaday.com/2017/09/26/an-introduction-to-solid-state-relays/) was helpful summing it up.

Probably for power regulation the Robotdyn (with direct control of the Triac) is the most elegant solution (even if running quite hot). And if the SSR keeps cooler, it may be an alternative for the bypass. In any case, I plan to use it for space heating ;-)

**Comment by JAndrassy** [comment]
I checked the schematics of the Grove Electricity Sensor (the CT module) and I wonder that it didn't destroy any analog pin with -5 V. Uno WiFi had it wired more then a year. I wired it to Wemos, M0, MKR. The Grove I2C ADC has some level shifting circuit on input which maybe protects it.
The schematics of the CT module was not available before and the description is misleading.

I ordered the new Grove AC sensor (<5A ACS70331). With it the MKR Grove Base and a new relay again. I replaced the bypass relay some weeks ago, now the main relay causes W5500 reset at AC disconnection while only the 40 W circulation pump is running.

**Comment by JAndrassy** [comment]
I could test the new Grove AC sensor today. It reads nice sine wave with AVR and with SAMD analogRead. The middle value is around 1.5 V so it is good for a 3.3 V MCU. 

I tested it with the MKR connector base too, but the shield has signal level shifting. For analog pins there is a voltage divider and a capacitor. The capacitor smoothed the AC wave to no wave.

**Comment by stealth-ultd** [comment]
Would it work if you connect it directly to an analog pin on the MKR Zero, or otherwise through the I2C ADC board to the MKR connector carrier?

**Comment by JAndrassy** [comment]
> Would it work if you connect it directly to an analog pin on the MKR Zero, or otherwise through the I2C ADC board to the MKR connector carrier?

yes I tested it then directly connected and it works perfect with 3.3 V Vcc and ADC.

**Comment by stealth-ultd** [comment]
If the sensor proves to work when directly connected to an analog pin on the MKR Zero, the capacitor on the shield could be taken out (soldering it off). It's a bit fiddly but possible.

When your load hits 2kW, what does the sensor read? The specs state that it can handle 10A, but maxes out measuring at 5A, which should be somewhere around 1.2kW. Does it still provide reliable readings beyond that?

**Comment by JAndrassy** [comment]
analogRead is close to zero at - 5 A. at zero crossing analogRead reads around 450 on MKR.

Yesterday I tested only with a 40 W bulb and with a 1 kW portable heater. In the Regulator box I will split the AC line and measure only half of the current. See the updated schematics in README

**Comment by JAndrassy** [comment]
update

the EMI problem with main relay was not solved with a new relay, but with swap to pin 3. 

the Grove AC sensor is usable. less current goes over it with the split wire the I expected, so the read values are only +-100, but the `readElSens` function returns good values. (I multiple the float with 1000 for the return value.) and the `elsensPower` calculation with modified coefficients gives accurate values.

If we have some sunny day at weekend I will try a 30 A hall effect AC sensor without AC wire split to compare with the Grove AC sensor readings.

**Comment by stealth-ultd** [comment]
Meanwhile, I have ordered all the components (some arrive from Chine, so it will still take some time). Different temperature sensor (MCP9808, I2C module). For the current sensor I have settled for a ACS712-based 20A module in a neat DIN-rail version. Bit over-the-top, but it seems well made.

Appreciate the recent updates made to the repository! It makes it all much more straightforward to implement. Fronius technical support has confirmed that neither the inverter Datacard, nor MPPT has any "insight" in the surplus available under zero-export. No surprise, but good to have confirmed.

**Comment by JAndrassy** [comment]
Yes, the current sensor I still plan to test is ACS712 30 A. The sensor is 5 V and has the middle output value at 2.5 V. So for a 3.3 V board the measured current can generate only up to 3.3 V. For your 1500 W the 20 A version should be good and for my 2000 W the 30 A version should not go over 3.3 V at 9 A.

and about DIN rail mount: did you see this? https://industruino.com/page/products

**Comment by stealth-ultd** [comment]
Had not seen those modules before. If the Regulator goes into series production it may be a good option ;-)

These current sensors should work indeed. And with the MKR Connector Carrier (& DAC) it could use the entire range of the sensor, as far as I can see. "All the I/Os are supplied at 5V and the carrier provides the proper 5V to 3.3V level shifting.".

I am considering to extrapolate the "surplus" using a separate mini solar panel (with load resister & DAC) next to the main PV panels, to measure the amount of irradiation (or alternatively, % of output). With that info the surplus could be approximated. Might be worth giving a try...

**Comment by JAndrassy** [comment]
All the analog connectors of the Carrier have voltage divider and a capacitor and the capacitor smooths the AC wave away. I don't know what for is the capacitor and how do this connectors work as output.

an LDR could measure the sun light too (protected by sun glasses :-) )

**Comment by JAndrassy** [comment]
so finally I use the 30 A ACS712. the splitted wire for the 5 V sensor distorts the measurement.

the new commit with the code modifications for ACS712 contains Calc sheets I use for coefficients of formulas and readme.md with comments. the sheets have example data

**Comment by JAndrassy** [comment]
@stealth-ultd, did you build it? Or you bought the Ohmpilot?

**Comment by stealth-ultd** [comment]
Sorry for the delayed response. It is here on the bench - no Ohmpilot in sight! But before connecting (and deciding on the necessary changes to be implemented in the code) I am waiting for a change in the inverter setup. Likely, we will change from 1-phase to 3-phase setup and have export enabled. Need permission from network regulator for this. Just everything is slowed down now because of lockdown. If we get permission, it will stay closer to your setup.

**Comment by stealth-ultd** [comment]
It has been a while since I have been coding, so taking this Regulator for a new start has been “interesting”. That said, the code is laid out elegantly and pretty well-annotated, so it is advancing good. The Regulator is since a few days feeding a storage heater (6.4 kWh thermal battery) from excess PV supplemented by nighttime tariff “imported” electricity.

We are still waiting for permission for the Distributive Grid Operator (takes months, unfortunately) to connect a three-phase inverter and allow export. So, for now, our regulator implementation remains somewhat improvised and regulated through MQTT commands from Node-Red rather than ModBus. Not running a pump meant I could recycle its alarm code for MQTT client connection (alike the network alarm function). I opted for a somewhat lower “beep” tone for the beeper, though ;-) Piece by piece, it is coming together – and yes, the modularity of the code is very helpful in this.

Yesterday got ElSens working (with ACS712 20A through MKR Carrier, with A0 resistor unsoldered), however, I am now a bit puzzled with the calibration of the readings.

- In principle, I am using a “perfect” 0.8kW resistive load, so the Power factor should be nearly 1. What settings should be used for PF_ANGLE_INTERVAL and PF_ANGLE_SHIFT in this case (and how did you figure the values for your system, where I guess the main 2kW heating element is also nearly “perfect” resistive apart from the 40W pump)?
- There is a scaling factor ELSENS_VALUE_COEF = 1.86 – is this number based on reverse-engineering the value to get a matching elsensPower value, or else?

**Comment by JAndrassy** [comment]
I am glad my project helps.

I write about ElSens here https://github.com/jandrassy/Regulator/tree/master/util/Calc#elsenspower

it is not a 'true' power factor like when the AC voltage and current are shifted. this 'power factor' is from cutting away part of the current wave. 

I expolored the EMI from relay. I switch the relays at zero crossing, but It turns out that the bypass relay at full load takes 10 milliseconds until it disconnects so it disconnects at next zero crossing. The main relay which at switch off has only 40 W of the circulation pump, switches off after 5 milliseconds so exactly at voltage peek which creates EMI.

**Comment by stealth-ultd** [comment]
Thanks for the quick response! I'll start with manually adding some samples to RegulatorDev.ods and see where we get with calibrating Elsens. Right now I am using still a different thyristor library (dimmable_light_linearized.h), plan to switch to TriacLib in the next days.

I have been using a solid-state relay with ESP8622 on which the code was running before, but the MKR Zero has a slightly lower output voltage on the pin and it won't switch the SSR for bypass mode (yes indeed, that dimmer gets hot, even with an extra heatsink attached!). Thinking about driving the SSR from a mosfet. When I get it all sorted I will hook on the oscilloscope and check how clean the output stays. But first things first ...

**Comment by stealth-ultd** [comment]
The Calc part is definitively helpful. I've now also implemented the TriacLib, calibrated power to period with dimmer-dev.ods. Manually calibrating with RegulatorDev.ods works like a charm.

**Comment by stealth-ultd** [comment]
A quick question as I am about to implement the Modbus functionality. A look at the code tells that the inverter needs to be in slave mode ("Inverter control via Modbus") and that you write to the interface (modbus.write). Is this required to read out the data? I have no battery functionality in the Galvo (and upcoming Symo) so I am not intentding to change the state of the inverter through the Regulator.

**Comment by JAndrassy** [comment]
Regulator uses Modbus TCP. Modbus TCP has server and client, not master and slave. The Symo is server and Regulator is a client. 
Modbus RTU can't be used because there Data Manager is master and Smart Meter and BMS are slaves. Ohmpilot is slave too and the Data Manager firmware as master sends data to Ohmpilot as slave. There can be only one master in Modbus RTU.

**Comment by stealth-ultd** [comment]
I am aware of this, I guess my question was not clear.
In the web admin of the inverterou set to control the inverter via Modbus. Why is that "control" necessary? OR: is it still necessary if I just want to read (request) data from the inverter and smart meter via Modbus TCP for the Regulator, without any intention to influence the actual state of the inverter. It is that "control" bit that confuses me here.

**Comment by JAndrassy** [comment]
sorry. yes 'control' is to allow for write operations. I use the modbusWriteSingle() function in Modbus.ino only from BattSett.ino

**Comment by stealth-ultd** [comment]
Clear, thanks :-)

BTW: Is the symo service password crucial for any of this communication? I see it used in SusCalib.ino but maybe that's not relevant for my setup (no battery).

**Comment by JAndrassy** [comment]
SusCalib uses JSON API (an undocumented request used by the web GUI).
modbus tcp doesn't require a password

**Comment by stealth-ultd** [comment]
Modbus working for a day now – stable, no issues. Compliments to the code. I figured that "int + SF" had to get selected in inverter settings for things to work.

**Comment by stealth-ultd** [comment]
Since a week, insolation (W/m2) is getting measured with a small reference solar panel. The 6V, 1W panel has a max shortcut current of 0.19A, which over the 15 Ohms resistor produces ~3V at max solar power. Output is being measured with a Grove 12-bit ADC module and is referenced against the real-world output of the inverter. Now we can finally “regulate”, although the snowfall disrupts the procedure! Beta-testing will be going on for a bit before all is set and calibrated.

One quirk is that the triac seems to be randomly firing when the period is set to zero, at times (unpredictable?) giving bursts up to 1/3 of max output for some time. I will report about this separately.

![IMG_TA-Pilot-small](https://user-images.githubusercontent.com/47171697/104608961-ce313280-5682-11eb-9cb2-0127e3ea3704.jpg) 

Components used (from left-up to right-down):
(row 1) solid state relay (SSR), dimmer module with extra heatsink, cable connector block
(row 2) MOSFET driver for SSR, I2C hub, DIN-mount current sensor, OLED display, LED-bar, DIN-mount power supply
(row 3) MKR Zero/ETH on MKR carrier, speaker, ADC for reference mini PV panel, hacked voltage divider (now connector) for PV panel, 15 OHM 5W load resistor for mini PV panel

**Comment by JAndrassy** [comment]
thank you for the picture. I see you added a tiny small I2C display. the red module is the main 'El.sens.' hall sensor? the white DIN component is the main relay? and what is the gray DIN component?

**Comment by stealth-ultd** [comment]
(I have now added the components listing under the picture) The I2C OLED is helpful for debugging. The red module is a MOSFET driver for the main SSR relay module (white), as the MKR does not provide enough voltage to switch the SSR. Grey DIN component is a Meanwell 12V, 1.25A power supply – it is very compact.

**Comment by JAndrassy** [comment]
I thought the CT is the power supply :-). 
The SSR is the main relay? Grove has a MOSFET module. I used it for PWM 3.3 to 5 V for the Kemo module.

In my box I replaced the power supply for a small one with 5 V output. See photo

**Reply by stealth-ultd** [reply]
I did not implement the main relay. The SSR is the bypass relay, in parallel with the Triac module. It stays nicely cool.

**Comment by stealth-ultd** [comment]
There is an issue with the Triac seemingly randomly firing smaller or larger bursts of power. They are short “blips” or longer periods of “stutter”, followed by shorter or longer periods without the issue appearing. I got aware of it when elsens surprised me with reading power, while setPeriod is zero and state = monitoring. To be sure it is not just a sensor ghost reading, I attached a light bulb, confirming the flicker.

I have tried compiling with Arduino SAMD21 CPU Core 1.8.9  and 1.8.11 with no difference. I disconnected the Oled display (maybe a bad I2C implementation?) and switched on / off chunks of sketch code (MQTT causing disruption?, etc.) – but the incidental random firing remains. One thing: I noticed that the line voltage reported by the Fronius smart meter varies quite a bit – sometimes from 218 to 228 within a few minutes. I confirmed this with a true-rms voltmeter. So the power line sees quite a bit of variation and probably noise.

In the discussions of another Triac library (https://github.com/fabiuz7/Dimmable-Light-Arduino) the possibility of power line noise interfering with the Triac is mentioned: "If you encounter flickering problem due to noise on electrical network, you can try to enable (uncommenting) FILTER_INT_PERIOD defined at the begin of thyristor.cpp file.” Maybe it provides a clue to what we see with TriacLib / Robotdyn?

((off-topic: this library allows control of the effective delivered power, not just thyristor's activation time, which sounds like a neat bit of functionality))

So not sure what is at play: a “dirty” interrupt lock, or a “dirty” zero-crossing detection, or something else?

**Comment by JAndrassy** [comment]
that is the problem. no main relay. sometimes I see in data a small rise of current while the power is set to zero and the main relay is still on, but after the main relay is off I have no problem :-). so I didn't implement full off in TriacLib.
the TriacLib is always 'armed' starting one shot 'PWM' at zero crossing. for 0 power the pulse is just before next zero crossing. it can happen that the pulse runs over the zero crossing and the triac is on for next half wave.

one solution would be to set shorter pulse and/or limiting the value at the zero power end as it is limited at the full power end:
https://github.com/jandrassy/Regulator/blob/d32090828f2bfae284e7a95e3b55631ca64f6198/TriacLib/TriacLib.h#L199-L202

or the zero crossing interrupt could be turned off with a dedicated function. then there would be no pulse to triac.

**Reply by stealth-ultd** [reply]
Ok, now I get it. The main relay essentially "unplugs" the regulator ;-)
But would you consider the last option (introducing a dedicated "off" function) the most elegant way to go?

**Reply by JAndrassy** [reply]
> Ok, now I get it. The main relay essentially "unplugs" the regulator ;-)
> But would you consider the last option (introducing a dedicated "off" function) the most elegant way to go?

yes. should be easy. I try to do it today

EDIT: sorry, I didn't get to it. I had problems to solve with [current projects hardware](https://github.com/jandrassy/SunriseClock/blob/main/README.md#images), I wouldn't be able to concentrate on TriacLib.

**Reply by stealth-ultd** [reply]
OK :-)
I meanwhile also had an unexpected issue: sun melted from the small reference PV panel, but the "real" solar array has still lots of snow on it, so I had to make some fixes to prevent the regulator get too enthusiastic too early with switching on power!

**Reply by stealth-ultd** [reply]
I had a look at the TriacLib code yesterday but honestly have no idea where to start with this. Appreciate if you could give it a try, if it's not too big of a thing.

**Comment by JAndrassy** [comment]
Hello

<s>it is done. TriacLib has `stop()` and `start()` for SAMD. 
`setup()` **doesn't** call start().
start() and stop() connect/disconnect the event system between interrupt and timer, so Triac::waitZeroCrossing will always work.
I tested with with a light bulb. I will integrate it in PowerPilot later. But I guess you have your version of PowerPilot without batteries support and using the input from the 'solar sensor'.</s>

I have surplus power first in March. Then I plan to add to PowerPilot support for zero export setup if I_STATUS_THROTTLED will be reported over Modbus as expected.

Juraj

**Comment by stealth-ultd** [comment]
Thanks for this! I will try to see how best to implement.

((BTW: In the new PowerPilot the Triac:start(); seems to miss "::" [as in Triac**::**start();] if I understand well?))

In general, I am still reviewing how my setup works (it looks all OK, but had little sun for robust testing!). Once all is fine, I can share my implementation, if you are interested. Currently trying to implement a weather-prediction based night-charging module.

**Comment by JAndrassy** [comment]
<s>I fixed the ::. thank you. I decided to remove start() from setup() in last moment before commit and I had to add it in PowerPilot to not to forget about it, but I didn't verify the Regulator project then (the TriacLib test is a separate sketch which I tested after the change of course).</s>

If you make a git fork of my Regulator repository, then it will be possible to compare and move commits between the forks.

**Comment by JAndrassy** [comment]
**sorry, I changed it, but it is simpler now. start and stop is handled in setPeriod. I removed the start() and stop() functions**

**Comment by stealth-ultd** [comment]
I am running it since this morning and all looks perfect, thanks! This now saves me from hacking the PowerPilot to make the start/stop work :-)

**Comment by stealth-ultd** [comment]
Latest state of hardware, with among others addition of a bigger display (Oled SSD1327, 128 x 128).

![TA-Pilot_2-small](https://user-images.githubusercontent.com/47171697/105754347-3cf77100-5f4a-11eb-9fb0-a5c6132d7798.jpg)

Components used (from left-up to right-down):
(row 1) solid state relay (SSR), dimmer module with extra heatsink, cable connector block
(row 2) MOSFET driver for SSR, I2C hub, DIN-current sensor, LED-bar, DIN-power supply, DIN-ETH overvoltage protection
(row 3) MKR Zero/ETH on MKR carrier, speaker, OLED display, ADC for reference mini PV panel, 15 OHM 5W load resistor for mini PV panel

**Comment by stealth-ultd** [comment]
I am implementing WebServer.ino. Two quick questions.
(1) webServerLoop() uses the msg command ("msg.print" etc.) but when I compile, it complains ('msg' was not declared in this scope). Till now I had always commented it out everywhere else in the sketch, but how/where is this function actually defined?
(2) Running the webpages static from laptop now works OK. But from the SD card not ("/index~1.htm" not found). My first thought is that the pages are maybe not in the good location – do they need to be simply placed in the root of the SD folder?

Some hints appreciated :-)

**Reply by JAndrassy** [reply]
1) msg is global in Regulator.ino https://github.com/jandrassy/Regulator/blob/5cdb4be8ba940ec28b74401a7d94ce3f234ee274/Regulator/Regulator.ino#L79
and is printed by Telnet.ino. 

2) yes in the root of the SD card

**Reply by stealth-ultd** [reply]
Thanks for getting back to this.
1. This is clear (msg) – I'll leave it commented out as I am not using Telnet right now.
2. Still struggling with the webpages. While accessing locally works OK (from /data folder on computer), through the web only partly. What works: I can access and download the CSV file (DATA2021.CSV). What does not work: I cannot access the pages. For instance: index.html returns "/index~1.htm" not found – the address pointer seems to get modified according to the code (line 116 - 122 WebServer.ino). I don't get yet what this piece of code is for, and how I can resolve this issue.

**Reply by JAndrassy** [reply]
I didn't want to have the files in data folder with .htm extension. But on FAT file system of the SD card they are stored shortened to 8.3 file name. So I do a simple conversion of filename.html to 8.3 version filena~1.htm

**Comment by stealth-ultd** [comment]
Heating season is on again, so the Regulator has been started up! Plan to do some modifications and fixes. Will report progress (and fork the code when it's sorted out).

**Comment by zooz124** [comment]
@stealth-ultd Hi, do you have a repository of what you have done until now? I have a zero export installation too, and I am interested only on dimming some heaters with the surplus energy. A simple one without extra features. Thanks in advance.

**Reply by stealth-ultd** [reply]
@zooz124 Hi, I will do some cleanup of ongoing experiments and post it in a day or two.

**Comment by stealth-ultd** [comment]
Finally – I have now forked the zero-export branch. Was about time :-)

**Reply by JAndrassy** [reply]
I [compared](https://github.com/jandrassy/Regulator/compare/master...stealth-ultd:zero-export) with your fork. I very like how you reused my Regulator code base.

**Reply by stealth-ultd** [reply]
Glad to hear. I appreciate how you have set up the code (modular, well documented). Very helpful to add or change functionality!

**Comment by JAndrassy** [comment]
@stealth-ultd I simplified the power2period formula. It makes more sense now and is faster to calculate (the speed doesn't matter much though).
It is now `POWER2PERIOD_SHIFT + POWER2PERIOD_KOEF * acos(1 - 2 * ratio)` and the POWER2PERIOD_KOEF should be half of the old value (because it included division by PI and now it is 2*PI)

I modified dimmer-dev.ods, power2period.ods and util/README.md too. The new 'ratio' line in chart in dimmer-dev.ods shows the input values for the formula.

**Comment by JAndrassy** [comment]
I replaced the speaker module with Grove passive buzzer and I adjusted the beeper frequency to the buzzer's optimal frequency. It is loud when it beeps and completely silent in between. It doesn't catch interference.

**Comment by stealth-ultd** [comment]
Some months ago I updated some of the libraries used in this sketch, and since the webserver output gives an error (the index page loads, shows the 'Values' label, but for the rest remains blank with a pop-up window telling just '0'. Checked the code multiple times but found nothing obvious. Today I quickly reverted back to StreamLib 1.2.1 and it all works perfectly again. Updating to 1.3.0 gives a compiler error (on MKR Zero) and 1.3.1 breaks the web page. Not sure if this is an issue with StreamLib, or how WebServer.ino or script.js handle the request, or something else. In any case, I thought better to report it.

**Reply by JAndrassy** [reply]
Hello. The only difference I see compared to my version is that I didn't run the project with Ethernet library with StreamLib 1.3.1. 
I had EthernetENC with ATmega, then I switched to WiFi still with the ATmega in July. In August and September I run the project on SAMD boards MKR 1000 (on Carier) and then Nano 33 IoT (it is SAMD too).
I will test the project with MKR and Ethernet library.

**Reply by JAndrassy** [reply]
I can't replicate your problem.
I have one small fix in WebServer.ino, but even if I revert it everything works.
The small fix was size-1 for readBytesUntil to reserve space for terminating 0 if some unexpected long line is read from client.
```
int l = client.readBytesUntil(' ', fn, sizeof(fn) - 1);
```

**Comment by stealth-ultd** [comment]
Hello, thanks for looking into this! I am out of reach of the Regulator right now, but after the weekend I'll try out the small fix to see if it does the trick.

**Comment by JAndrassy** [comment]
Hello. I recommend you changes from commit https://github.com/JAndrassy/Regulator/commit/3fbee9ea17160b8bdf93a4cc11525c571e226b56
